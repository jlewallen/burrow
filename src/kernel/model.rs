use anyhow::Result;
use nanoid::nanoid;
use once_cell::sync::Lazy;
use serde::{Deserialize, Serialize};
use std::{
    cell::RefCell,
    collections::HashMap,
    fmt::{Debug, Display},
    ops::{Deref, DerefMut, Index},
    rc::{Rc, Weak},
};
use thiserror::Error;
use tracing::{debug, span, trace, Level};

use super::{infra::*, Scope};

pub static WORLD_KEY: Lazy<EntityKey> = Lazy::new(|| EntityKey("world".to_string()));

pub static NAME_PROPERTY: &str = "name";

pub static DESC_PROPERTY: &str = "desc";

pub static GID_PROPERTY: &str = "gid";

#[derive(Clone)]
pub struct EntityPtr {
    entity: Rc<RefCell<Entity>>,
    lazy: LazyLoadedEntity,
}

impl EntityPtr {
    pub fn new_blank() -> Self {
        let brand_new = Rc::new(RefCell::new(Entity::default()));
        let lazy = LazyLoadedEntity::new_from_raw(&brand_new);

        Self {
            entity: brand_new,
            lazy,
        }
    }

    pub fn downgrade(&self) -> Weak<RefCell<Entity>> {
        Rc::downgrade(&self.entity)
    }

    pub fn key(&self) -> EntityKey {
        self.lazy.key.clone()
    }
}

impl From<Rc<RefCell<Entity>>> for EntityPtr {
    fn from(ep: Rc<RefCell<Entity>>) -> Self {
        let lazy = LazyLoadedEntity::new_from_raw(&ep);

        Self {
            entity: Rc::clone(&ep),
            lazy,
        }
    }
}

impl From<Entity> for EntityPtr {
    fn from(entity: Entity) -> Self {
        Rc::new(RefCell::new(entity)).into()
    }
}

impl Debug for EntityPtr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("EntityPtr")
            .field("key", &self.lazy.key)
            .field("name", &self.lazy.name)
            .finish()
    }
}

// This seems cleaner than implementing borrow/borrow_mut ourselves and things
// were gnarly when I tried implementing Borrow<T> myself.
impl Deref for EntityPtr {
    type Target = RefCell<Entity>;

    fn deref(&self) -> &Self::Target {
        self.entity.as_ref()
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Serialize, Deserialize, Debug)]
pub struct EntityGID(i64);

impl EntityGID {
    pub fn new(i: i64) -> EntityGID {
        EntityGID(i)
    }

    pub fn gid_to_string(&self) -> String {
        format!("{}", self.0)
    }
}

impl Display for EntityGID {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl From<EntityGID> for i64 {
    fn from(gid: EntityGID) -> Self {
        gid.0
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
pub struct EntityKey(String);

impl EntityKey {
    pub fn new(s: &str) -> EntityKey {
        EntityKey(s.to_string())
    }

    pub fn key_to_string(&self) -> &str {
        &self.0
    }
}

impl Debug for EntityKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("`{}`", self.0))
    }
}

impl Default for EntityKey {
    fn default() -> Self {
        Self(nanoid!())
    }
}

impl Display for EntityKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

pub trait DomainEvent: Debug {}

#[derive(Debug)]
pub enum DomainOutcome {
    Ok(Vec<Box<dyn DomainEvent>>),
    Nope,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Item {
    Named(String),
    Route(String),
    GID(EntityGID),
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct EntityRef {
    #[serde(rename = "py/object")]
    py_object: String,
    #[serde(rename = "py/ref")]
    py_ref: String,
    pub key: EntityKey,
    #[serde(rename = "klass")]
    class: String,
    name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Identity {
    #[serde(rename = "py/object")]
    py_object: String,
    private: String,
    public: String,
    signature: Option<String>, // TODO Why does this happen in the model?
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Kind {
    #[serde(rename = "py/object")]
    py_object: String,
    identity: Identity,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct EntityClass {
    #[serde(rename = "py/type")]
    py_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AclRule {
    #[serde(rename = "py/object")]
    py_object: String,
    keys: Vec<String>,
    perm: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Acls {
    #[serde(rename = "py/object")]
    py_object: String,
    rules: Vec<AclRule>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Version {
    #[serde(rename = "py/object")]
    py_object: String,
    i: u32,
}

impl Default for Version {
    fn default() -> Self {
        Self {
            py_object: Default::default(),
            i: 1,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Property {
    #[serde(rename = "py/object")]
    py_object: String,
    acls: Acls,
    value: serde_json::Value,
}

impl Property {
    pub fn new(value: serde_json::Value) -> Self {
        Self {
            py_object: "".to_string(),
            acls: Default::default(),
            value,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Props {
    #[serde(rename = "py/object")]
    py_object: String,
    map: HashMap<String, Property>,
}

impl Default for Props {
    fn default() -> Self {
        Self {
            py_object: "model.properties.Common".to_string(), // #python-class
            map: Default::default(),
        }
    }
}

impl Props {
    fn property_named(&self, name: &str) -> Option<&Property> {
        if self.map.contains_key(name) {
            return Some(self.map.index(name));
        }
        None
    }

    fn string_property(&self, name: &str) -> Option<String> {
        if let Some(property) = self.property_named(name) {
            match &property.value {
                serde_json::Value::String(v) => Some(v.to_string()),
                _ => None,
            }
        } else {
            None
        }
    }

    fn i64_property(&self, name: &str) -> Option<i64> {
        if let Some(property) = self.property_named(name) {
            match &property.value {
                serde_json::Value::Number(v) => v.as_i64(),
                _ => None,
            }
        } else {
            None
        }
    }

    fn set_property(&mut self, name: &str, value: serde_json::Value) {
        self.map.insert(name.to_string(), Property::new(value));
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Entity {
    #[serde(rename = "py/object")]
    py_object: String,
    pub key: EntityKey,
    version: Version,
    parent: Option<LazyLoadedEntity>,
    creator: Option<LazyLoadedEntity>,
    identity: Identity,
    #[serde(rename = "klass")]
    class: EntityClass,
    acls: Acls,
    props: Props,
    scopes: HashMap<String, serde_json::Value>,

    #[serde(skip)] // Very private
    infra: Option<Weak<dyn Infrastructure>>,
}

impl Display for Entity {
    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
        fmt.debug_struct(&self.class.py_type)
            .field("key", &self.key)
            .field("name", &self.name())
            .field("gid", &self.gid())
            .finish()
    }
}

impl Needs<std::rc::Rc<dyn Infrastructure>> for Entity {
    fn supply(&mut self, infra: &std::rc::Rc<dyn Infrastructure>) -> Result<()> {
        self.infra = Some(Rc::downgrade(infra));
        self.parent = infra.ensure_optional_entity(&self.parent)?;
        self.creator = infra.ensure_optional_entity(&self.creator)?;
        Ok(())
    }
}

impl Entity {
    pub fn name(&self) -> Option<String> {
        self.props.string_property(NAME_PROPERTY)
    }

    pub fn set_name(&mut self, value: &str) -> Result<()> {
        let value: serde_json::Value = value.into();
        self.props.set_property(NAME_PROPERTY, value);

        Ok(())
    }

    pub fn desc(&self) -> Option<String> {
        self.props.string_property(DESC_PROPERTY)
    }

    pub fn gid(&self) -> Option<EntityGID> {
        if let Some(id) = self.props.i64_property(GID_PROPERTY) {
            Some(EntityGID(id))
        } else {
            None
        }
    }

    pub fn has_scope<T: Scope>(&self) -> bool {
        self.scopes.contains_key(<T as Scope>::scope_key())
    }

    pub fn scope_mut<T: Scope>(&mut self) -> Result<OpenScopeMut<T>, DomainError> {
        let scope = self.load_scope::<T>()?;
        Ok(OpenScopeMut::new(self, scope))
    }

    pub fn scope<T: Scope>(&self) -> Result<OpenScope<T>, DomainError> {
        let scope = self.load_scope::<T>()?;
        Ok(OpenScope::new(scope))
    }

    pub fn maybe_scope<T: Scope>(&self) -> Result<Option<OpenScope<T>>, DomainError> {
        if !self.has_scope::<T>() {
            return Ok(None);
        }
        let scope = self.load_scope::<T>()?;
        Ok(Some(OpenScope::new(scope)))
    }

    fn load_scope<T: Scope>(&self) -> Result<Box<T>, DomainError> {
        let scope_key = <T as Scope>::scope_key();

        let _load_scope_span = span!(
            Level::DEBUG,
            "scope",
            key = self.key.key_to_string(),
            scope = scope_key
        )
        .entered();

        if !self.scopes.contains_key(scope_key) {
            return Ok(Box::new(Default::default()));
        }

        // The call to serde_json::from_value requires owned data and we have a
        // reference to somebody else's. Presumuably so that we don't couple the
        // lifetime of the returned object to the lifetime of the data being
        // referenced? What's the right solution here?
        // Should the 'un-parsed' Scope also owned the parsed data?
        let data = &self.scopes[scope_key];
        let owned_value = data.clone();
        let mut scope: Box<T> = serde_json::from_value(owned_value)?;

        let _prepare_span = span!(Level::DEBUG, "prepare").entered();

        if let Some(infra) = &self.infra {
            if let Some(infra) = infra.upgrade() {
                scope.supply(&infra)?;
            } else {
                return Err(DomainError::ExpiredInfrastructure);
            }
        } else {
            return Err(DomainError::NoInfrastructure);
        }

        Ok(scope)
    }

    fn replace_scope<T: Scope>(&mut self, scope: &T) -> Result<()> {
        let scope_key = <T as Scope>::scope_key();

        let _span = span!(
            Level::DEBUG,
            "scope",
            key = self.key.key_to_string(),
            scope = scope_key
        )
        .entered();

        let value = scope.serialize()?;

        debug!("scope-replace");

        self.scopes.insert(scope_key.to_string(), value);

        Ok(())
    }
}

pub struct OpenScope<T: Scope> {
    target: Box<T>,
}

impl<T: Scope> OpenScope<T> {
    pub fn new(target: Box<T>) -> Self {
        trace!("scope-open {:?}", target);

        Self { target }
    }
}

impl<T: Scope> Deref for OpenScope<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.target.as_ref()
    }
}

pub struct OpenScopeMut<'me, T: Scope> {
    owner: &'me mut Entity,
    target: Box<T>,
}

impl<'me, T: Scope> OpenScopeMut<'me, T> {
    pub fn new(owner: &'me mut Entity, target: Box<T>) -> Self {
        trace!("scope-open {:?}", target);

        Self { owner, target }
    }

    pub fn save(&mut self) -> Result<()> {
        self.owner.replace_scope(self.target.as_ref())?;

        Ok(())
    }
}

impl<'me, T: Scope> Drop for OpenScopeMut<'me, T> {
    fn drop(&mut self) {
        // TODO Check for unsaved changes to this scope and possibly warn the
        // user, this would require them to intentionally discard  any unsaved
        // changes. Not being able to bubble an error up makes doing anything
        // elaborate in here a bad idea.
        trace!("scope-dropped {:?}", self.target);
    }
}

impl<'me, T: Scope> Deref for OpenScopeMut<'me, T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.target.as_ref()
    }
}

impl<'me, T: Scope> DerefMut for OpenScopeMut<'me, T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        self.target.as_mut()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct LazyLoadedEntity {
    #[serde(rename = "py/object")]
    py_object: String,
    #[serde(rename = "py/ref")]
    py_ref: String,
    pub key: EntityKey,
    #[serde(rename = "klass")]
    class: String,
    name: String,

    #[serde(skip)]
    entity: Option<Weak<RefCell<Entity>>>,
}

impl LazyLoadedEntity {
    pub fn new_with_entity(entity: EntityPtr) -> Self {
        Self::new_from_raw(&entity.entity)
    }

    fn new_from_raw(entity: &Rc<RefCell<Entity>>) -> Self {
        let shared_entity = entity.borrow();
        Self {
            py_object: "model.entity.EntityRef".to_string(), // #python-class
            py_ref: "model.entity.Entity".to_string(),       // #python-class
            key: shared_entity.key.clone(),
            class: shared_entity.class.py_type.clone(),
            name: shared_entity.name().unwrap_or_default(),
            entity: Some(Rc::downgrade(entity)),
        }
    }

    pub fn has_entity(&self) -> bool {
        self.entity.is_some()
    }

    pub fn into_entity(&self) -> Result<EntityPtr, DomainError> {
        match &self.entity {
            Some(e) => match e.upgrade() {
                Some(e) => Ok(e.into()),
                None => Err(DomainError::DanglingEntity),
            },
            None => Err(DomainError::DanglingEntity),
        }
    }
}

impl From<EntityPtr> for LazyLoadedEntity {
    fn from(entity: EntityPtr) -> Self {
        entity.lazy
    }
}

#[derive(Error, Debug)]
pub enum DomainError {
    #[error("no such scope '{:?}' on entity '{:?}'", .0, .1)]
    NoSuchScope(EntityKey, String),
    #[error("parse failed")]
    ParseFailed(#[source] serde_json::Error),
    #[error("dangling entity")]
    DanglingEntity,
    #[error("anyhow error")]
    Anyhow(#[source] anyhow::Error),
    #[error("no infrastructure")]
    NoInfrastructure,
    #[error("expired infrastructure")]
    ExpiredInfrastructure,
    #[error("session closed")]
    SessionClosed,
    #[error("container required")]
    ContainerRequired,
}

impl From<serde_json::Error> for DomainError {
    fn from(source: serde_json::Error) -> Self {
        DomainError::ParseFailed(source)
    }
}

impl From<anyhow::Error> for DomainError {
    fn from(source: anyhow::Error) -> Self {
        DomainError::Anyhow(source)
    }
}

#[derive(Error, Debug)]
pub enum EvaluationError {
    #[error("parse failed")]
    ParseFailed,
}

impl From<nom::Err<nom::error::Error<&str>>> for EvaluationError {
    fn from(source: nom::Err<nom::error::Error<&str>>) -> EvaluationError {
        match source {
            nom::Err::Incomplete(_) => EvaluationError::ParseFailed,
            nom::Err::Error(_e) => EvaluationError::ParseFailed,
            nom::Err::Failure(_e) => EvaluationError::ParseFailed,
        }
    }
}
