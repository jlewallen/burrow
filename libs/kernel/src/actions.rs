use anyhow::Result;
use serde_json::Value;
use std::{fmt::Debug, rc::Rc};

use crate::{Audience, DomainEvent, When};

use super::{session::SessionRef, Entry, Surroundings};

pub use replies::*;

pub type ReplyResult = anyhow::Result<Effect>;

/// TODO Make generic over SessionRef, Surroundings, and Result.
pub trait Action: ToJson + Debug {
    fn is_read_only() -> bool
    where
        Self: Sized;

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult;
}

#[derive(Debug, Clone)]
pub struct Raised {
    pub key: String,
    pub audience: Audience,
    pub event: Rc<dyn DomainEvent>,
}

impl Raised {
    pub fn new(audience: Audience, key: String, event: Rc<dyn DomainEvent>) -> Self {
        Self {
            key,
            audience,
            event,
        }
    }

    pub fn has_prefix(&self, prefix: &str) -> bool {
        self.key.starts_with(prefix)
    }
}

#[derive(Debug, Clone)]
pub struct Incoming {
    pub key: String,
    pub value: serde_json::Value,
}

impl Incoming {
    pub fn new(key: String, value: serde_json::Value) -> Self {
        Self { key, value }
    }

    pub fn has_prefix(&self, prefix: &str) -> bool {
        self.key.starts_with(prefix)
    }
}

#[derive(Clone, Debug)]
pub struct Scheduling {
    pub key: String,
    pub when: When,
    pub message: serde_json::Value,
}

#[derive(Clone, Debug)]
#[non_exhaustive]
pub enum Perform {
    Living {
        living: Entry,
        action: Rc<dyn Action>, // TODO Consider making this recursive?
    },
    Chain(Rc<dyn Action>),
    Delivery(Incoming),
    Raised(Raised),
    Schedule(Scheduling),
    Ping(TracePath),
}

pub trait Performer {
    fn perform(&self, perform: Perform) -> Result<Effect>;
}

#[derive(Clone, Debug)]
#[non_exhaustive]
pub enum RevertReason {
    Mysterious,
    Deliberate(String),
}

#[derive(Clone, Debug)]
#[non_exhaustive]
pub enum Effect {
    Ok,
    Prevented,
    Nothing,
    // Revert(RevertReason),
    Reply(Rc<dyn Reply>),
    Pong(TracePath),
    // This is tempting. Right now we recursively call the performer. I'm not
    // sure this gives us in benefit, but it could come in really handy when we
    // start to dynamically alter behavior in chained actions. Leaving this as
    // it stands until I have a stronger opinion.
    // Chain(Rc<dyn Action>),
}

#[derive(Clone, Default)]
pub struct TracePath(Vec<String>);

impl TracePath {
    pub fn push(self, v: String) -> Self {
        Self(self.0.into_iter().chain([v]).collect())
    }
}

impl Debug for TracePath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let value: String = self.clone().into();
        f.write_str(&value)
    }
}

impl Into<String> for TracePath {
    fn into(self) -> String {
        self.0.join("")
    }
}

impl ToJson for Effect {
    fn to_json(&self) -> std::result::Result<Value, serde_json::Error> {
        // TODO I'll need to work on this, if not to make tests scale.
        match self {
            Effect::Reply(reply) => reply.to_json(),
            _ => todo!(),
        }
    }
}

impl<T: Reply + 'static> From<T> for Effect {
    fn from(value: T) -> Self {
        Self::Reply(Rc::new(value))
    }
}
