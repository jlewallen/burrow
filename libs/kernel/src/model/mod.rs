use anyhow::Result;
use std::{
    cell::RefCell,
    collections::HashMap,
    fmt::Debug,
    ops::{Deref, Index},
    rc::Rc,
};

pub mod base;
pub mod builder;
pub mod compare;
pub mod entity;
pub mod entity_ref;
pub mod entry;
pub mod props;
pub mod scopes;

pub use base::*;
pub use builder::*;
pub use entity::*;
pub use entity_ref::*;
pub use entry::*;
pub use props::*;
pub use scopes::*;

#[derive(Clone)]
pub struct EntityPtr(Rc<RefCell<Entity>>);

impl EntityPtr {
    pub fn new(e: Entity) -> Self {
        Self(Rc::new(RefCell::new(e)))
    }
}

impl Deref for EntityPtr {
    type Target = RefCell<Entity>;

    fn deref(&self) -> &Self::Target {
        self.0.as_ref()
    }
}

pub trait IntoEntry {
    fn to_entry(&self) -> Result<Entry, DomainError>;
}

impl IntoEntry for EntityRef {
    fn to_entry(&self) -> Result<Entry, DomainError> {
        use super::session::get_my_session;
        if !self.key().valid() {
            return Err(DomainError::InvalidKey);
        }
        get_my_session()?
            .entry(&LookupBy::Key(self.key()))?
            .ok_or(DomainError::DanglingEntity)
    }
}

use compare::{AnyChanges, CompareChanges, CompareError, Modified, Original};

pub fn any_entity_changes(
    l: AnyChanges<Option<Original>, EntityPtr>,
) -> Result<Option<Modified>, CompareError> {
    use compare::TreeDiff;

    let value_after = {
        let entity = l.after.borrow();

        serde_json::to_value(&*entity)?
    };

    let value_before: JsonValue = if let Some(original) = &l.before {
        match original {
            Original::String(s) => s.parse()?,
            Original::Json(v) => (*v).clone(),
        }
    } else {
        JsonValue::Null
    };

    let diff = TreeDiff {};

    diff.any_changes(AnyChanges {
        before: value_before,
        after: value_after,
    })
}
