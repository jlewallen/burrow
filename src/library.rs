use nom::{bytes::complete::take_while1, combinator::map, IResult};

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Item {
    Described(String),
}

pub fn word(i: &str) -> IResult<&str, &str> {
    take_while1(move |c| "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".contains(c))(i)
}

pub fn spaces(i: &str) -> IResult<&str, &str> {
    take_while1(move |c| " \t".contains(c))(i)
}

pub fn noun(i: &str) -> IResult<&str, Item> {
    map(word, |s: &str| Item::Described(s.to_owned()))(i)
}
