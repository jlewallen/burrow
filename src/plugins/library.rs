pub mod parser {
    pub use crate::kernel::*;
    pub use nom::{
        branch::alt,
        bytes::complete::{tag, take_while1},
        character::complete::digit1,
        combinator::map,
        combinator::{map_res, recognize},
        sequence::{pair, preceded, separated_pair},
        IResult,
    };
    pub use tracing::*;

    pub fn word(i: &str) -> IResult<&str, &str> {
        take_while1(move |c| "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".contains(c))(i)
    }

    pub fn spaces(i: &str) -> IResult<&str, &str> {
        take_while1(move |c| " \t".contains(c))(i)
    }

    pub fn noun(i: &str) -> IResult<&str, Item> {
        map(word, |s: &str| Item::Named(s.to_owned()))(i)
    }

    pub fn unsigned_number(i: &str) -> IResult<&str, u64> {
        map_res(recognize(digit1), str::parse)(i)
    }

    pub fn gid_reference(i: &str) -> IResult<&str, Item> {
        map(preceded(tag("#"), unsigned_number), |n| {
            Item::GID(EntityGID::new(n))
        })(i)
    }

    pub fn noun_or_specific(i: &str) -> IResult<&str, Item> {
        alt((noun, gid_reference))(i)
    }

    pub fn named_place(i: &str) -> IResult<&str, Item> {
        alt((
            gid_reference,
            map(word, |s: &str| Item::Route(s.to_owned())),
        ))(i)
    }
}

pub mod model {
    pub use crate::kernel::*;
    pub use anyhow::Result;
    pub use serde::{Deserialize, Serialize};
    pub use serde_json::{json, Value};
    pub use std::{collections::HashMap, ops::Deref};
    pub use tracing::*;
}

pub mod actions {
    pub use crate::kernel::*;
    pub use crate::plugins::log_test;
    pub use crate::plugins::tools;
    pub use anyhow::Result;
    pub use tracing::*;
}
