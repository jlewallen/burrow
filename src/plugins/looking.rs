use crate::plugins::library::plugin::*;

#[derive(Default)]
pub struct LookingPlugin {}

impl Plugin for LookingPlugin {
    fn plugin_key() -> &'static str
    where
        Self: Sized,
    {
        "looking"
    }

    fn register_hooks(&self, _hooks: &crate::domain::ManagedHooks) {}
}

impl ParsesActions for LookingPlugin {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        try_parsing(parser::LookActionParser {}, i)
    }
}

pub mod model {
    use crate::plugins::library::model::*;
    use crate::{
        plugins::carrying::model::{Carryable, Containing},
        plugins::moving::model::{Movement, Occupyable},
    };

    pub fn qualify_name(quantity: f32, name: &str) -> String {
        use indefinite::*;
        use inflection::*;
        if quantity > 1.0 {
            let pluralized = plural::<_, String>(name);
            format!("{} {}", quantity, &pluralized)
        } else {
            indefinite(name)
        }
    }

    pub trait Observe<T> {
        fn observe(&self, user: &Entry) -> Result<T>;
    }

    impl Observe<ObservedEntity> for &Entry {
        fn observe(&self, _user: &Entry) -> Result<ObservedEntity> {
            let name = self.name()?;
            let carryable = self.scope::<Carryable>()?;
            let qualified = name.as_ref().map(|n| qualify_name(carryable.quantity(), n));

            Ok(ObservedEntity {
                key: self.key.to_string(),
                name,
                qualified,
                desc: self.desc()?,
            })
        }
    }

    pub fn new_inside_observation(user: &Entry, vessel: &Entry) -> Result<InsideObservation> {
        let mut items = vec![];
        if let Ok(containing) = vessel.scope::<Containing>() {
            for lazy_entity in &containing.holding {
                let entity = &lazy_entity.into_entry()?;
                items.push(entity.observe(user)?);
            }
        }

        Ok(InsideObservation {
            vessel: vessel.observe(user)?,
            items,
        })
    }

    pub fn new_area_observation(user: &Entry, area: &Entry) -> Result<AreaObservation> {
        // I feel like there's a lot of unnecessary copying going on here.

        let mut living: Vec<ObservedEntity> = vec![];
        if let Ok(occupyable) = area.scope::<Occupyable>() {
            for entity in &occupyable.occupied {
                living.push((&entity.into_entry()?).observe(user)?);
            }
        }

        let mut items = vec![];
        if let Ok(containing) = area.scope::<Containing>() {
            for entity in &containing.holding {
                items.push((&entity.into_entry()?).observe(user)?);
            }
        }

        let mut carrying = vec![];
        if let Ok(containing) = user.scope::<Containing>() {
            for entity in &containing.holding {
                carrying.push((&entity.into_entry()?).observe(user)?);
            }
        }

        let mut routes = vec![];
        if let Ok(movement) = user.scope::<Movement>() {
            for route in &movement.routes {
                routes.push((&route.area.into_entry()?).observe(user)?);
            }
        };

        Ok(AreaObservation {
            area: area.observe(user)?,
            person: user.observe(user)?,
            living,
            items,
            carrying,
            routes,
        })
    }
}

pub mod actions {
    use super::model::*;
    use crate::plugins::library::actions::*;

    #[derive(Debug)]
    pub struct LookAction {}

    impl Action for LookAction {
        fn is_read_only() -> bool {
            true
        }

        fn perform(&self, args: ActionArgs) -> ReplyResult {
            info!("look!");

            let (_, user, area, _) = args.unpack();

            Ok(Box::new(new_area_observation(&user, &area)?))
        }
    }

    #[derive(Debug)]
    pub struct LookInsideAction {
        pub item: Item,
    }

    impl Action for LookInsideAction {
        fn is_read_only() -> bool {
            true
        }

        fn perform(&self, args: ActionArgs) -> ReplyResult {
            info!("look inside!");

            let (_, user, _area, infra) = args.unpack();

            match infra.find_item(&args, &self.item)? {
                Some(target) => {
                    if tools::is_container(&target)? {
                        Ok(Box::new(new_inside_observation(&user, &target)?))
                    } else {
                        Ok(Box::new(SimpleReply::Impossible))
                    }
                }
                None => Ok(Box::new(SimpleReply::NotFound)),
            }
        }
    }
}

pub mod parser {
    use crate::plugins::library::parser::*;

    use super::actions::{LookAction, LookInsideAction};

    pub struct LookActionParser {}

    impl ParsesActions for LookActionParser {
        fn try_parse_action(&self, i: &str) -> EvaluationResult {
            let inside = map(
                separated_pair(
                    separated_pair(tag("look"), spaces, tag("inside")),
                    spaces,
                    noun,
                ),
                |(_, nearby)| Box::new(LookInsideAction { item: nearby }) as Box<dyn Action>,
            );

            let area = map(tag("look"), |_| Box::new(LookAction {}) as Box<dyn Action>);

            let (_, action) = alt((inside, area))(i)?;

            Ok(action)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::model::*;
    use super::parser::LookActionParser;
    use super::*;
    use crate::{
        domain::{BuildActionArgs, QuickThing},
        plugins::library::plugin::try_parsing,
    };

    #[test]
    fn it_looks_in_empty_area() -> Result<()> {
        let mut build = BuildActionArgs::new()?;
        let args: ActionArgs = build.plain().try_into()?;

        let action = try_parsing(LookActionParser {}, "look")?;
        let reply = action.perform(args.clone())?;
        let (_, _person, _area, _) = args.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_looks_in_area_with_items_on_ground() -> Result<()> {
        let mut build = BuildActionArgs::new()?;
        let args: ActionArgs = build
            .ground(vec![QuickThing::Object("Cool Rake")])
            .ground(vec![QuickThing::Object("Boring Shovel")])
            .try_into()?;

        let action = try_parsing(LookActionParser {}, "look")?;
        let reply = action.perform(args.clone())?;
        let (_, _person, _area, _) = args.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_looks_in_area_with_items_on_ground_and_a_route() -> Result<()> {
        let mut build = BuildActionArgs::new()?;
        let destination = build.make(QuickThing::Place("Place"))?;
        let args: ActionArgs = build
            .ground(vec![QuickThing::Object("Cool Rake")])
            .ground(vec![QuickThing::Object("Boring Shovel")])
            .route("East Exit", QuickThing::Actual(destination))
            .try_into()?;

        let action = try_parsing(LookActionParser {}, "look")?;
        let reply = action.perform(args.clone())?;
        let (_, _person, _area, _) = args.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_looks_in_area_with_items_on_ground_and_holding_items() -> Result<()> {
        let mut build = BuildActionArgs::new()?;
        let destination = build.make(QuickThing::Place("Place"))?;
        let args: ActionArgs = build
            .ground(vec![QuickThing::Object("Boring Shovel")])
            .hands(vec![QuickThing::Object("Cool Rake")])
            .route("East Exit", QuickThing::Actual(destination))
            .try_into()?;

        let action = try_parsing(LookActionParser {}, "look")?;
        let reply = action.perform(args.clone())?;
        let (_, _person, _area, _) = args.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_fails_to_look_inside_non_containers() -> Result<()> {
        let mut build = BuildActionArgs::new()?;
        let args: ActionArgs = build
            .hands(vec![QuickThing::Object("Not A Box")])
            .try_into()?;

        let action = try_parsing(LookActionParser {}, "look inside box")?;
        let reply = action.perform(args.clone())?;
        let (_, _person, _area, _) = args.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn it_looks_inside_containers() -> Result<()> {
        let mut build = BuildActionArgs::new()?;
        let vessel = build
            .build()?
            .named("Vessel")?
            .holding(&vec![build.make(QuickThing::Object("Key"))?])?
            .into_entry()?;
        let args: ActionArgs = build.hands(vec![QuickThing::Actual(vessel)]).try_into()?;

        let action = try_parsing(LookActionParser {}, "look inside vessel")?;
        let reply = action.perform(args.clone())?;
        let (_, _person, _area, _) = args.unpack();

        insta::assert_json_snapshot!(reply.to_json()?);

        build.close()?;

        Ok(())
    }

    #[test]
    fn qualify_name_basics() {
        // Not going to test all of indefinite's behavior here, just build edge
        // cases in our integrating logic.
        assert_eq!(qualify_name(1.0, "box"), "a box");
        assert_eq!(qualify_name(2.0, "box"), "2 boxes");
        assert_eq!(qualify_name(1.0, "person"), "a person");
        assert_eq!(qualify_name(2.0, "person"), "2 people");
        assert_eq!(qualify_name(1.0, "orange"), "an orange");
        assert_eq!(qualify_name(2.0, "orange"), "2 oranges");
        assert_eq!(qualify_name(1.0, "East Exit"), "an East Exit");
    }
}
