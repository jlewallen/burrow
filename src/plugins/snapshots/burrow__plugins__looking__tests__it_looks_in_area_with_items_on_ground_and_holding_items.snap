---
source: src/plugins/looking.rs
expression: reply.to_json()?
---
{
  "areaObservation": {
    "area": {
      "desc": null,
      "key": "E-4",
      "name": "Welcome Area",
      "qualified": "a Welcome Area"
    },
    "carrying": [
      {
        "desc": null,
        "key": "E-3",
        "name": "Cool Rake",
        "qualified": "a Cool Rake"
      }
    ],
    "items": [
      {
        "desc": null,
        "key": "E-5",
        "name": "Boring Shovel",
        "qualified": "a Boring Shovel"
      },
      {
        "desc": null,
        "key": "E-6",
        "name": "East Exit",
        "qualified": "an East Exit"
      }
    ],
    "living": [
      {
        "desc": null,
        "key": "E-2",
        "name": "Living",
        "qualified": "a Living"
      }
    ],
    "person": {
      "desc": null,
      "key": "E-2",
      "name": "Living",
      "qualified": "a Living"
    },
    "routes": []
  }
}
